### PR TITLE
[IMP] account: payment method improvements

### PR DESCRIPTION
**Pull request by vin-odoo**
To improve the payment method system, proceed to a few changes
such as changing the view a bit, making sure payment acquirers are not
linked to a journal by default and that only the manual payment method
type can be used multiple times in a single journal.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_move.py [hunk 1]
```diff
@@ -1497,7 +1497,7 @@ def _get_reconciled_info_JSON_values(self):
                 'payment_id': counterpart_line.id,
                 'partial_id': partial.id,
                 'account_payment_id': counterpart_line.payment_id.id,
-                'payment_method_name': counterpart_line.payment_id.payment_method_id.name if counterpart_line.journal_id.type == 'bank' else None,
+                'payment_method_name': counterpart_line.payment_id.payment_method_line_id.name if counterpart_line.journal_id.type == 'bank' else None,
```
Comment (smetl) [inline addons/account/models/account_move.py:1500]: `counterpart_line.payment_id.payment_method_line_id.name` is enough. A payment can only be on a bank journal :)

File: addons/account/tests/test_account_journal.py [hunk 2]
```diff
@@ -131,3 +108,31 @@ def _get_payment_method_information(self):
...
+            'amount': 100.0,
+            'payment_type': 'inbound',
+            'partner_type': 'customer',
+            'payment_method_line_id': first_method.id,
+        })
+
+        first_method.unlink()
+
+        self.assertTrue(first_method.exists())
+        self.assertEqual(first_method.journal_id.id, False)
```
Comment (smetl) [inline addons/account/tests/test_account_journal.py:132]: shorter: `self.assertFalse(first_method.journal_id)`

File: addons/payment/models/account_journal.py [hunk 3]
```diff
@@ -30,3 +30,8 @@ def _check_inbound_payment_method_line_ids(self):
         if acquirers:
             raise UserError(_("You can't delete a payment method that is linked to an acquirer in the enabled or test state.\n"
                               "Linked acquirer(s): %s", ', '.join(a.display_name for a in acquirers)))
+
+    def _get_available_payment_method_lines(self, payment_type):
+        lines = super(AccountJournal, self)._get_available_payment_method_lines(payment_type)
```
Comment (smetl) [inline addons/payment/models/account_journal.py:35]: shorter: `lines = super()._get_available_payment_method_lines(payment_type)`

File: addons/payment/models/account_payment_method.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        related='payment_acquirer_id.state'
+    )
+
+    @api.depends('payment_method_id')
+    def _compute_payment_acquirer_id(self):
+        for line in self:
+            code = line.payment_method_id.code
+            line.payment_acquirer_id = self.env['payment.acquirer'].search([
+                ('provider', '=', code), ('company_id', '=', line.journal_id.company_id.id)
+            ], limit=1)
```
Comment (smetl) [inline addons/payment/models/account_payment_method.py:25]: `search` in a loop is not efficient. Why not something like:
```
@api.depends('payment_method_id')
def _compute_payment_acquirer_id(self):
    acquirers = self.env['payment.acquirer'].search([
        ('provider', 'in', self.mapped('code')),
        ('company_id', 'in', self.journal_id.company_id.id),
    ])
    acquirers_map = {(x.provider, x.company_id): x for x in acquirers}
    for line in self:
        code = line.payment_method_id.code
        company = line.journal_id.company_id
        line.payment_acquirer_id = acquirers_map.get((code, company), False)
```

File: addons/payment/models/account_payment_method.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    def _search_payment_acquirer_id(self, operator, operand):
+        return [('id', operator, operand)]
+
+    def action_open_acquirer_form(self):
+        return {
```
Comment (smetl) [inline addons/payment/models/account_payment_method.py:31]: missing `self.ensure_one()`

**Review by Feyensv — CHANGES_REQUESTED** [review]
Why do we even freeze versions ?

Part of this PR doesn't belong in stable (especially field changes).

**Comment by smetl** [comment]
@Feyensv We don't have the choice, dude! #fp/avw-request-we-need-this-asap-in-odoo.com

**Comment by tde-banana-odoo** [comment]
> @Feyensv We don't have the choice, dude! #fp/avw-request-we-need-this-asap-in-odoo.com

Antony said we don't have to do things because someone asked to. Challenge everything !

#iwillmergeitbutnotwithoutyellingatbigramsandinsultingalotofpeopleandsayingshitincommitmessage

File: addons/account/models/account_payment.py [hunk 6]
```diff
@@ -381,40 +378,42 @@ def _compute_is_internal_transfer(self):
...
-                pay.payment_method_id = available_payment_methods[0]._origin
+                pay.payment_method_line_id = available_payment_methods[0]._origin
             else:
-                pay.payment_method_id = False
+                pay.payment_method_line_id = False
 
     @api.depends('payment_type',
                  'journal_id.inbound_payment_method_line_ids',
                  'journal_id.outbound_payment_method_line_ids')
-    def _compute_payment_method_fields(self):
```
Comment (tivisse) [inline addons/account/models/account_payment.py:405]: FYI, removing a method doesn't respect the stable policy. 

See: https://github.com/odoo/odoo/wiki/Contributing#what-does-stable-mean

File: addons/payment/models/account_payment_method.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+from odoo import api, fields, models, _
+
+
+class AccountPaymentMethodLine(models.Model):
+    _inherit = "account.payment.method.line"
+
+    payment_acquirer_id = fields.Many2one(
+        comodel_name='payment.acquirer',
+        compute='_compute_payment_acquirer_id',
+        store=True
```
Comment (tivisse) [inline addons/payment/models/account_payment_method.py:13]: Modifying the DB scheme too.

**Comment by Feyensv** [comment]
> @Feyensv We don't have the choice, dude! #fp/avw-request-we-need-this-asap-in-odoo.com

You just have to send them anyone from upgrade/internal to clear their heads :thinking:.
I'm not even sure you will manage to convince runbot & upgrade team for the multiple exceptions needed before merge :P

**Comment by smetl** [comment]
Enough trolling guys^^

**Comment by tivisse** [comment]
@smetl I didn't see the other comments before posting it actually.

The truth is that if 3 people tell that this is not valid in a stable release, maybe they're right.

**Comment by tde-banana-odoo** [comment]
> Enough trolling guys^^

Actually it is not only trolling. If we want to have real freeze and move towards rolling releases, FP and AVW have to accept that some changes are not acceptable in stable. If you really want to do your change because otherwise some business flows are broken, you should do it using a custom module that you merge afterwards into base module in master. And also find why a broken flow was merged and reverted a few weeks afterwards.

This is indeed not my functional scope, but I think all guru should be somehow involved in maintaining stable as stable as possible when possible. And we all know it is hard to contain some bigrams or their mitosis.

**Comment by smetl** [comment]
@tivisse I already know about that. Consider this task as an exception. If you want to discuss deeper about that, I can put you into a meeting with the ones taking such decision. Let me know if you have time to waste ;)

**Comment by Xavier-Do** [comment]
robodoo p=0

**Pull request closed without merging**